### PR TITLE
Create Docker image for Node.js + Yarn

### LINKED ISSUE #1577: [Question] Official Docker image Yarn+Nodejs
Hey,

is there an official supported docker image with yarn and nodejs preinstalled that i can use for CI?

### LINKED ISSUE #2686: Need official node-yarn docker image
One of the biggest obstacles to my team adopting yarn was putting together a docker image that included yarn. After doing a bunch of work i've determined that this very easy but non-obvious. It doesn't help that search results return a number of individual contributions that are flawed in one way or another. Many use the npm install method which is not recommended and others use `curl -o- -L https://yarnpkg.com/install.sh | bash` which installs only for the current user.

Here's my take - https://hub.docker.com/r/mattkime/node-yarn/

Let me know if there's anything i can do to help move this forward.

### PR DESCRIPTION
**Pull request by Daniel15**
**Summary**
Creates a Docker image for Node.js and Yarn. This is just a small addition to the official Node.js Docker image that a lot of users could just do in their own Dockerfiles, we've received a few requests for it though.

I've already pushed this to the Docker hub as `yarnpkg/node-yarn`: https://hub.docker.com/r/yarnpkg/node-yarn/

**Test plan**
```
C:\src\yarn (docker-1) (yarn@0.22.0-0)
λ docker run -it --rm yarnpkg/node-yarn /bin/bash
root@13ead647da12:/# yarn --version
0.20.3
root@13ead647da12:/# cd /tmp
root@13ead647da12:/tmp# mkdir yarntest
root@13ead647da12:/tmp# cd yarntest
root@13ead647da12:/tmp/yarntest# yarn init
yarn init v0.20.3
question name (yarntest):
question version (1.0.0):
question description:
question entry point (index.js):
question repository url:
question author:
question license (MIT):
success Saved package.json
Done in 1.69s.
root@13ead647da12:/tmp/yarntest# yarn add react
yarn add v0.20.3
info No lockfile found.
[1/4] Resolving packages...
[2/4] Fetching packages...
[3/4] Linking dependencies...
[4/4] Building fresh packages...
success Saved lockfile.
success Saved 16 new dependencies.
├─ asap@2.0.5
├─ core-js@1.2.7
├─ encoding@0.1.12
├─ fbjs@0.8.9
├─ iconv-lite@0.4.15
├─ is-stream@1.1.0
├─ isomorphic-fetch@2.2.1
├─ js-tokens@3.0.1
├─ loose-envify@1.3.1
├─ node-fetch@1.6.3
├─ object-assign@4.1.1
├─ promise@7.1.1
├─ react@15.4.2
├─ setimmediate@1.0.5
├─ ua-parser-js@0.7.12
└─ whatwg-fetch@2.0.2
Done in 2.81s.
```

Closes #2686 
Closes #1577

### CODE REVIEW

File: Dockerfile.node7 [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# Dockerfile for using Yarn along with Node.js 7.x
+# docker build -t yarnpkg/node-yarn:latest -f Dockerfile.node7 .
```
Comment (SimenB) [inline Dockerfile.node7]: If you do `--squash` as well, there is no need for chained commands, the resulting image is just a single layer on top of `FROM`

**Comment by SimenB** [comment]
Could you add a image for node 6 as well? 😄  Current LTS, and 7 will be EOL in a couple of month

File: Dockerfile.node7 (hunk 1 above)
Comment (Daniel15) [inline Dockerfile.node7]: Does that make it slower to build the image, as the intermediate layers aren't cached? Not that there's many layers in this Dockerfile anyways 😛

**Comment by Daniel15** [comment]
Sure, I can do that. I should also add a script to rebuild all the images.

**Comment by Daniel15** [comment]
Filed a separate issue for it: #2737

File: Dockerfile.node7 (hunk 1 above)
Comment (SimenB) [inline Dockerfile.node7]: No, you'll still have all the layers locally, it's just the resulting image that's a single layer :smile:

**Comment by airtonix** [comment]
Should be using alpine-node instead of just the node image... size difference is huge, which matters in CI environments.

**Comment by Daniel15** [comment]
Yarn is included as part of the Node.js Docker images now, so you can use the Alpine image if you like. 

A lot of people use Debian packages as part of their build environment, so Alpine isn't always the best choice. There's not as much software packaged for Alpine compared to Debian.

**Comment by airtonix** [comment]
Trust me... You might think you want all those non-nodejs things to create
the mother of all docker images... But you'd be wrong.

**Comment by Daniel15** [comment]
Use the best tool for the job... Not everything can be solved with Node.js when stable solutions exist elsewhere 😃  For Yarn, we have a few extra non-Node.js utilities used as part of the build/release process (particularly for building Debian/RPM packages)

**Pull request merged**